app/data_models/data_stores.py

```diff
@@ -0,0 +1,21 @@
+from dataclasses import dataclass, field
+from typing import MutableMapping
+
+from app.data_models.answer_store import AnswerStore
+from app.data_models.list_store import ListStore
+from app.data_models.metadata_proxy import MetadataProxy
+from app.data_models.progress_store import ProgressStore
+from app.data_models.supplementary_data_store import SupplementaryDataStore
+
+
+@dataclass
+class DataStores:
```

Comment:
What break if you made this frozen?

Reply:
```
app/data_models/questionnaire_store.py:58: error: Property "metadata" defined in "DataStores" is read-only  [misc]
app/data_models/questionnaire_store.py:80: error: Property "supplementary_data_store" defined in "DataStores" is read-only  [misc]
app/data_models/questionnaire_store.py:143: error: Property "progress_store" defined in "DataStores" is read-only  [misc]
app/data_models/questionnaire_store.py:145: error: Property "supplementary_data_store" defined in "DataStores" is read-only  [misc]
app/data_models/questionnaire_store.py:148: error: Property "answer_store" defined in "DataStores" is read-only  [misc]
app/data_models/questionnaire_store.py:149: error: Property "list_store" defined in "DataStores" is read-only  [misc]
app/data_models/questionnaire_store.py:150: error: Property "response_metadata" defined in "DataStores" is read-only  [misc]
app/data_models/questionnaire_store.py:166: error: Property "response_metadata" defined in "DataStores" is read-only  [misc]

```
and multitude of tests, but the fixing above will lead to another host of issues somewhere else, from what I recall (when I tried it).

Reply:
That means, tests are written badly. What breaks outside of tests, what functionally breaks?.

Reply:
From what I recall, for starters, all calls to metadata throughout the code, for example:
```python
schema_name := metadata.schema_name
```
or
```python
contextvars.bind_contextvars(
        tx_id=metadata.tx_id,
        ce_id=metadata.collection_exercise_sid,
    )
```
will be broken since making dataclass frozen won't allow us to access metadata this way anymore (but from what I remember, accessing by key like `metadata["id"]` will cause another host of issues). I didn't dig much further as even fixing metadata and all the tests was a lot of churn and looking somehow beyond the scope of the card.

Reply:
Are key access like `metadata["id"]` disallowed when frozen, even though they are not being set?.

Reply:
I deleted the last comment, until I finished investigating some more, but basically, a lot of issues with the metedata, the fact that it is in both `questionnaire_store` and `data_stores`, cannot overcome some obstacles when initiating DataStore whenever it needs to be re-instantiated. Here's the original comment:




> From what I recall, for starters, all calls to metadata throughout the code, for example:
```python
schema_name := metadata.schema_name
```
> or
```python
contextvars.bind_contextvars(
        tx_id=metadata.tx_id,
        ce_id=metadata.collection_exercise_sid,
    )
```
> will be broken since making dataclass frozen won't allow us to access metadata this way anymore (but from what I remember, accessing by key like `metadata["id"]` will cause another host of issues). I didn't dig much further as even fixing metadata everywhere and all the tests was a lot of churn and looking somehow beyond the scope of the card. We had a discussion about making datastore just a separate, mutable class for now (on the same day) and I didn't pursue the idea of making it frozen any further and lost the changes since then.

Reply:
There is a lot of extra work making it work as frozen datastore, I am not sure if I should fully dedicate my time investigating it.

Reply:
Alright, let's leave it for now, there is enough indication that it isn't straight forward, so we can tackle it separately.

Reply:
Yeah, from what I checked so far, It needs some considerable time allocated for investigating, thinking through and implementing (it sort of branches off from this as a separate task).